File: tool/WDAExample.py
import sys

import whodunargs

class TestProgram(whodunargs.StandardProgram):
	def __init__(self):
		whodunargs.StandardProgram.__init__(self)
		self.name = "WDAExample"
		self.summary = "A simple example for whodunargs."
		self.usage = "python3 WDAExample"
		self.version = "WDAExample 0.0\nCopyright (C) 2023 Benjamin Crysup\nLicense LGPLv3: GNU LGPL version 3\nThis is free software: you are free to change and redistribute it.\nThere is NO WARRANTY, to the extent permitted by law.\n"
		# a flag option
		self.flagOpt = whodunargs.ArgumentOptionFlag("--bye","Test Flag","Test Flag")
		self.options.append(self.flagOpt)
		# an enum option
		self.enumVal = whodunargs.ArgumentOptionEnumValue("radio")
		self.enumOptA = whodunargs.ArgumentOptionEnum("--doa", "Enum A", "The first enum value.", self.enumVal)
		self.options.append(self.enumOptA)
		self.enumOptB = whodunargs.ArgumentOptionEnum("--dob", "Enum B", "The second enum value.", self.enumVal)
		self.options.append(self.enumOptB)
		self.enumOptC = whodunargs.ArgumentOptionEnum("--doc", "Enum C", "The third enum value.", self.enumVal)
		self.options.append(self.enumOptC)
		# an integer option
		self.intOpt = whodunargs.ArgumentOptionInteger("--qual","Quality","A simple integer option.","--qual 20")
		self.intOpt.value = 20
		self.options.append(self.intOpt)
		# a float option
		self.fltOpt = whodunargs.ArgumentOptionFloat("--pow","Power","A simple float option.","--pow 1.0")
		self.fltOpt.value = 1.0
		self.options.append(self.fltOpt)
		# a string option
		self.strOpt = whodunargs.ArgumentOptionString("--name","Name","A simple string option.","--name abc")
		self.strOpt.value = "abc"
		self.options.append(self.strOpt)
		# an integer vector option
		self.intvOpt = whodunargs.ArgumentOptionIntegerVector("--dims","Dimensions","A simple integer vector option.","--dims 10")
		self.options.append(self.intvOpt)
		# a float vector option
		self.fltvOpt = whodunargs.ArgumentOptionFloatVector("--cols","Collisions","A simple float vector option.","--cols 0.4")
		self.options.append(self.fltvOpt)
		# a string vector option
		self.strvOpt = whodunargs.ArgumentOptionStringVector("--tags","Tags","A simple string vector option.","--tags bad")
		self.options.append(self.strvOpt)
		# a file read option
		self.filerOpt = whodunargs.ArgumentOptionFileRead("--in","Input","A file read option.","--in /home/ben/in.tsv")
		self.filerOpt.validExts.add(".tsv")
		self.options.append(self.filerOpt)
		# a file write option
		self.filewOpt = whodunargs.ArgumentOptionFileWrite("--out","Output","A file write option.","--out /home/ben/in.csv")
		self.filewOpt.validExts.add(".csv")
		self.options.append(self.filewOpt)
		# a folder read option
		self.foldrOpt = whodunargs.ArgumentOptionFolderRead("--builddir","Build Directory","A folder read option.","--builddir /home/ben/")
		self.options.append(self.foldrOpt)
		# a folder write option
		self.foldwOpt = whodunargs.ArgumentOptionFolderWrite("--workd","Working Directory","A folder write option.","--workd /home/ben/")
		self.options.append(self.foldwOpt)
		# many file reads
		self.filervOpt = whodunargs.ArgumentOptionFileReadVector("--refs","References","A file read vector option.","--refs /home/ben/in.bmp")
		self.options.append(self.filervOpt)
		# many file writes
		self.filewvOpt = whodunargs.ArgumentOptionFileWriteVector("--logs","Log Output","A file write vector option.","--logs /home/ben/in.midi")
		self.options.append(self.filewvOpt)
	def baseRun(self):
		self.useOut.write(bytes("Hello\n","utf-8"))
		if self.flagOpt.value:
			self.useOut.write(bytes("Goodbye\n","utf-8"))


toRun = TestProgram()
toRun.parse(sys.argv[1:])
toRun.run()

File: ports/R/WDAExample.R

source("whodunargs.R")

TestProgram = R6Class("TestProgram",
	inherit = StandardProgram,
	public = list(
		flagO = NA,
		enumV = NA,
		enumOA = NA,
		enumOB = NA,
		intO = NA,
		fltO = NA,
		strO = NA,
		intvO = NA,
		fltvO = NA,
		strvO = NA,
		initialize = function(){
			super$initialize("TestProg", "A simple test program", "Use it.", "(C) ???")
			self$flagO = ArgumentOptionFlag$new("Test Flag", "--flag", "A flag option to test parsing.")
				self$addArgument(self$flagO)
			self$enumV = ArgumentOptionEnumValue$new("Test Enum")
			self$enumOA = ArgumentOptionEnum$new("Test Enum 1","--enumA","The first enum test.",self$enumV)
				self$addArgument(self$enumOA)
			self$enumOB = ArgumentOptionEnum$new("Test Enum 2","--enumB","The second enum test.",self$enumV)
				self$addArgument(self$enumOB)
			self$intO = ArgumentOptionInteger$new("Test Int","--int","An integer test","--int 8")
				self$addArgument(self$intO)
			self$fltO = ArgumentOptionFloat$new("Test Float","--float","A float test","--float 8.6")
				self$addArgument(self$fltO)
			self$strO = ArgumentOptionString$new("Test String","--string","A string test","--string abc")
				self$addArgument(self$strO)
			self$intvO = ArgumentOptionIntegerVector$new("Test Int Vector","--intvec","An integer vector test","--intvec 8")
				self$addArgument(self$intvO)
			self$fltvO = ArgumentOptionFloatVector$new("Test Float Vector","--floatvec","A float vector test","--floatvec 8.6")
				self$addArgument(self$fltvO)
			self$strvO = ArgumentOptionStringVector$new("Test String Vector","--stringvec","A string vector test","--stringvec abc")
				self$addArgument(self$strvO)
		},
		baseRun = function(){
			if(self$flagO$value){
				print("Flagged")
			}
			if(self$enumOA$value()){
				print("E 1")
			} else {
				print("E 2")
			}
			print(self$intO$value)
			print(self$fltO$value)
			print(self$strO$value)
			print(self$intvO$value)
			print(self$fltvO$value)
			print(self$strvO$value)
			invisible(self)
		}
	)
)

comArgs = commandArgs(trailingOnly=TRUE)
testP = TestProgram$new()
testP$parse(comArgs)$run()


